restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/SmsService.java

```diff
@@ -206,25 +213,41 @@ private void message(final Object message) throws IOException {
             if(logger.isDebugEnabled()) {
                 logger.debug("toOrganizationSid" + toOrganizationSid);
             }
-            Client toClient = clients.getClient(toUser, toOrganizationSid);
+            final Client toClient = clients.getClient(toUser, toOrganizationSid);
             if (toClient != null) { // looks like its a p2p attempt between two valid registered clients, lets redirect
-                // to the b2bua
-                if (B2BUAHelper.redirectToB2BUA(request, client, toClient, storage, sipFactory, patchForNatB2BUASessions)) {
-                    // if all goes well with proxying the SIP MESSAGE on to the target client
-                    // then we can end further processing of this request and send response to sender
-                    if(logger.isInfoEnabled()) {
-                        logger.info("P2P, Message from: " + client.getLogin() + " redirected to registered client: "
-                            + toClient.getLogin());
-                    }
-                    monitoringService.tell(new TextMessage(((SipURI)request.getFrom().getURI()).getUser(), ((SipURI)request.getTo().getURI()).getUser(), TextMessage.SmsState.INBOUND_TO_CLIENT), self);
-                    return;
-                } else {
-                    String errMsg = "Cannot Connect to Client: " + toClient.getFriendlyName()
-                            + " : Make sure the Client exist or is registered with Restcomm";
-                    sendNotification(errMsg, 11001, "warning", true);
-                    final SipServletResponse resp = request.createResponse(SC_NOT_FOUND, "Cannot complete P2P messages");
-                    resp.send();
+                long delay = pushNotificationServerHelper.sendPushNotificationIfNeeded(toClient.getPushClientIdentity());
+                // workaround for only clients with push_client_identity after long discussion about current SIP Message flow processing
+                // https://telestax.atlassian.net/browse/RESTCOMM-1159
+                if (delay > 0) {
+                    final SipServletResponse trying = request.createResponse(SipServletResponse.SC_TRYING);
+                    trying.send();
                 }
+                system.scheduler().scheduleOnce(Duration.create(delay, TimeUnit.MILLISECONDS), new Runnable() {
```

Comment:
this tasl logic is mostly sipservlet API related, this would be better handled in SIPServlet thread pool by using https://mobicents.ci.cloudbees.com/job/RestcommSipServlets-Release/lastSuccessfulBuild/artifact/documentation/jsr289-extensions-apidocs/org/mobicents/javax/servlet/sip/SipApplicationSessionAsynchronousWork.html

Reply:
The main motivation here is to use akka ecosystem. The same thing was implemented in CallManager and we agreed with @gvagenas about that

Reply:
well, this is new dicussion with @gvagenas . There is a pending change in RC to use SIPServlets from proper SipServletsASyncWork, so no concurrent situations lead to inconsistencies, but agree this is a different topic, up to you on how to do this...

Reply:
Let's keep it with akka scheduler just to leave this functionality with the same approach in all parts of RC. After that we can make a decision to use akka ecosystem or SipServletsASyncWork and to create new task for refactoring in whole RC code